### DISCUSSION TITLE
How to populate a field's `defaultValue` based on another collection?

### DISCUSSION THREAD

**Discussion by ViktorShev** [payloadcms/payload#9162] [General]
Hello, I've been trying to figure this out for a couple of days now and I'm kind of running out of ideas, so wanted to ask if anybody has done this before.

My use case is simple:

I have the following data structures:
- `Client`
- `Project` (has a relationship to `Client`)
- `Estimate` (has a relationship to `Project`, has many `Milestone`s)
- `Milestone` (has many `Epic`s)
- `Epic`

`Client` has a field called `uncertaintyMultipliers` which is an object of the structure: `{ low: number, medium: number, high: number, extreme: number }`

and `Project` has the same thing, but they act as overrides: `uncertaintyMultipliersOverrides`. (A project's uncertainty multipliers override the client multipliers).

My issue arises here (checkout the TODO comment):

```
export const Epics: CollectionConfig = {
  access: {
    create: () => true,
    delete: () => true,
    read: () => true,
    update: () => true,
  },
  admin: {
    useAsTitle: 'name',
  },
  fields: [
    {
      label: 'Name',
      name: 'name',
      required: true,
      type: 'text',
    },
    {
      label: 'Description',
      name: 'description',
      type: 'textarea',
    },
    {
      label: 'Phase',
      name: 'phase',
      type: 'text',
    },
   ...
    {
      defaultValue: 1.1, // TODO: This should be a reference to the client's uncertainty multipliers or project if there is an override
      label: 'Uncertainty',
      name: 'uncertaintyMultiplier',
      required: true,
      type: 'number',
    },
  ...
  ],
  slug: 'epics',
}
```

What I'm trying to achieve is basically: an Epic's DEFAULT uncertainty multiplier should be the "Low" multiplier of either the project or the client, depending if it's being overriden or not, think of it this way: 

`defaultValue: project.uncertaintyMultipliers.low ?? client.unceratintyMultipliers.low`

I have figured out that `defaultValue` can be a function even though it is not typed as such, and that it receives one argument which is an object containing the current user.

I have tried many ways of getting my hands on the Epic's ID to then to make a search for the project it belongs to and get the configured multipliers for the respective project or client and then set the default value to the `low` multiplier, but so far I could not figure it out, is there a way to do this?

The current thing I'm trying is using a custom component for that field and doing the fetching there, since I can access the Epic's ID through the URL, but honestly, it's not a solution I'm too content with.

It would be AMAZING if the `defaultValue` property would receive information about the current item being viewed, like its ID, since Payload is already aware of it when fetching the item's data from the DB to display it on screen.

Thanks in advance.

Versions:

```
"@payloadcms/db-mongodb": "3.0.0-beta.117",
"@payloadcms/db-sqlite": "3.0.0-beta.117",
"@payloadcms/graphql": "3.0.0-beta.117",
"@payloadcms/live-preview-react": "3.0.0-beta.117",
"@payloadcms/plugin-nested-docs": "3.0.0-beta.117",
"@payloadcms/plugin-redirects": "3.0.0-beta.117",
"@payloadcms/plugin-seo": "3.0.0-beta.117",
"@payloadcms/richtext-lexical": "3.0.0-beta.117",
"@payloadcms/ui": "3.0.0-beta.117",
"next": "^15.0.0",
"payload": "3.0.0-beta.117",
"react": "^19.0.0-rc-65a56d0e-20241020",
"react-dom": "^19.0.0-rc-65a56d0e-20241020",
```

**Comment by akhrarovsaid** [comment]
Hey @ViktorShev,

So the way I would handle this would be via a [Custom Field Component](https://payloadcms.com/docs/beta/admin/fields#custom-components). You would fetch the related collections' multipliers and the set the value as needed, and you would do this work prior to the user being able to edit (which makes me believe that you would have a field structure such that the initial component is a RSC, and then drill that fetched data down to a client component). A custom Text/Number field would do this no problem.

**Reply by ViktorShev** [reply]
Hey @akhrarovsaid, thanks for the quick reply.

Yeah this is the only solution I could come up with, I was just wondering if there's a more idiomatic/simpler way of achieving the same.

Unfortunately I don't think I can use a RSC to do the data fetching, since I need to access the URL to get the `Epic`'s ID, unless there's a hook that can provide me with it server-side (be it a payload or nextjs hook).

For now I'm stuck using a client component so I can read the ID (extracted from the URL) and doing data fetching the old `useEffect` way. This is why I'm not too content with this solution, because I could be using Payload instead of `fetch` + `useEffect` with type assertions.

**Reply by akhrarovsaid** [reply]
Hey @ViktorShev,

My pleasure. So keep in mind that custom components on the server are passed `params` in the props directly. You can type your field according to the docs (`PayloadServerComponent<CustomComponent<MyProps>>` or something along those lines), and in those `params` is an array: `segments` which I believe will contain the `id` of the entry, as well as the collection slug. This should allow you to use an RSC. Or, you could use a client component, either will be fine for this. I'm just thinking that there's some optimization to gain here from doing that work on the server.

EDIT: For fields, the data of the current doc is passed via a prop named `data`.

See second table here for [props passed to field server comps](https://payloadcms.com/docs/beta/admin/fields#default-props)

EDIT 2: And here for [typescript types for custom field comps](https://payloadcms.com/docs/beta/admin/fields#typescript)

**Reply by ViktorShev** [reply]
Just tested this, It seems like when I'm viewing the list of Epics it does give me a `data` field, `searchParams` and other stuff, but when I go to edit a specific item, I no longer get that information, I just get `payload` and some other stuff, but nothing related to the current document.

When viewing/editing a specific epic (/admin/collections/epics/someId):

```
{
    i18n: {
      dateFNS: [Object],
      dateFNSKey: 'en-US',
      fallbackLanguage: 'en',
      language: 'en',
      t: [Function: t],
      translations: [Object]
    },
    payload: BasePayload {
      auth: [AsyncFunction: auth],
      authStrategies: [Array],
      collections: [Object],
      config: [Object],
      count: [AsyncFunction: count],
      create: [AsyncFunction: create],
      db: [Object],
      decrypt: [Function: decrypt],
      duplicate: [AsyncFunction: duplicate],
      email: [Object],
      encrypt: [Function: encrypt],
      extensions: undefined,
      find: [AsyncFunction: find],
      findByID: [AsyncFunction: findByID],
      findGlobal: [AsyncFunction: findGlobal],
      findGlobalVersionByID: [AsyncFunction: findGlobalVersionByID],
      findGlobalVersions: [AsyncFunction: findGlobalVersions],
      findVersionByID: [AsyncFunction: findVersionByID],
      findVersions: [AsyncFunction: findVersions],
      forgotPassword: [AsyncFunction: forgotPassword],
      getAdminURL: [Function: getAdminURL],
      getAPIURL: [Function: getAPIURL],
      globals: [Object],
      importMap: [Object],
      logger: [EventEmitter],
      login: [AsyncFunction: login],
      resetPassword: [AsyncFunction: resetPassword],
      restoreGlobalVersion: [AsyncFunction: restoreGlobalVersion],
      restoreVersion: [AsyncFunction: restoreVersion],
      schema: undefined,
      secret: '224f432a43862a61cc76237f3c5a6dda',
      sendEmail: [AsyncFunction: sendEmail],
      types: undefined,
      unlock: [AsyncFunction: unlock],
      updateGlobal: [AsyncFunction: updateGlobal],
      validationRules: undefined,
      verifyEmail: [AsyncFunction: verifyEmail],
      versions: {}
    },
    clientField: {
      admin: [Object],
      name: 'uncertaintyMultiplier',
      type: 'number',
      label: 'Uncertainty Multiplier',
      virtual: false,
      _schemaPath: 'uncertaintyMultiplier'
    },
    field: {
      admin: [Object],
      name: 'uncertaintyMultiplier',
      type: 'number',
      label: 'Uncertainty Multiplier',
      validate: [Function (anonymous)],
      virtual: false,
      hooks: {},
      access: {}
    }
  }
```

When viewing the whole list (/admin/collections/epics):

 ```
  {
    collectionConfig: {
      access: [Object],
      admin: [Object],
      auth: false,
      custom: {},
      endpoints: [],
      fields: [Array],
      hooks: [Object],
      timestamps: true,
      upload: false,
      versions: false,
      slug: 'epics',
      labels: [Object],
      joins: {}
    },
    collectionSlug: 'epics',
    data: {
      docs: [Array],
      totalDocs: 7,
      limit: 10,
      totalPages: 1,
      page: 1,
      pagingCounter: 1,
      hasPrevPage: false,
      hasNextPage: false,
      prevPage: null,
      nextPage: null
    },
    hasCreatePermission: true,
    i18n: {
      dateFNS: [Object],
      dateFNSKey: 'en-US',
      fallbackLanguage: 'en',
      language: 'en',
      t: [Function: t],
      translations: [Object]
    },
    limit: 10,
    listPreferences: undefined,
    listSearchableFields: undefined,
    locale: undefined,
    newDocumentURL: '/admin/collections/epics/create',
    params: { segments: [Array] },
    payload: BasePayload {
      auth: [AsyncFunction: auth],
      authStrategies: [Array],
      collections: [Object],
      config: [Object],
      count: [AsyncFunction: count],
      create: [AsyncFunction: create],
      db: [Object],
      decrypt: [Function: decrypt],
      duplicate: [AsyncFunction: duplicate],
      email: [Object],
      encrypt: [Function: encrypt],
      extensions: undefined,
      find: [AsyncFunction: find],
      findByID: [AsyncFunction: findByID],
      findGlobal: [AsyncFunction: findGlobal],
      findGlobalVersionByID: [AsyncFunction: findGlobalVersionByID],
      findGlobalVersions: [AsyncFunction: findGlobalVersions],
      findVersionByID: [AsyncFunction: findVersionByID],
      findVersions: [AsyncFunction: findVersions],
      forgotPassword: [AsyncFunction: forgotPassword],
      getAdminURL: [Function: getAdminURL],
      getAPIURL: [Function: getAPIURL],
      globals: [Object],
      importMap: [Object],
      logger: [EventEmitter],
      login: [AsyncFunction: login],
      resetPassword: [AsyncFunction: resetPassword],
      restoreGlobalVersion: [AsyncFunction: restoreGlobalVersion],
      restoreVersion: [AsyncFunction: restoreVersion],
      schema: undefined,
      secret: '224f432a43862a61cc76237f3c5a6dda',
      sendEmail: [AsyncFunction: sendEmail],
      types: undefined,
      unlock: [AsyncFunction: unlock],
      updateGlobal: [AsyncFunction: updateGlobal],
      validationRules: undefined,
      verifyEmail: [AsyncFunction: verifyEmail],
      versions: {}
    },
    permissions: { canAccessAdmin: true, collections: [Object] },
    searchParams: { limit: '10' },
    user: {
      id: '671a97729facb4d8d9896b0a',
      role: 'admin',
      email: 'viktor@viktor.com',
      createdAt: '2024-10-24T18:52:34.824Z',
      updatedAt: '2024-10-25T12:40:37.155Z',
      loginAttempts: 0,
      collection: 'users',
      _strategy: 'local-jwt'
    },
    clientField: {
      admin: [Object],
      name: 'uncertaintyMultiplier',
      type: 'number',
      label: 'Uncertainty Multiplier',
      virtual: false,
      _schemaPath: 'uncertaintyMultiplier'
    },
    field: {
      admin: [Object],
      name: 'uncertaintyMultiplier',
      type: 'number',
      label: 'Uncertainty Multiplier',
      validate: [Function (anonymous)],
      virtual: false,
      hooks: {},
      access: {}
    }
  }
```
